row.py

```diff
@@ -988,9 +980,131 @@ def join_spreadsheet_info(filtered_results_file, out_dir):
+
+    #: add a 'keep' field, initialized as 'yes', but be used to mark rows to remove with 'no'
+    working_df["keep"] = "yes"
+
+    #: upper-case all letters
+    working_df["text"] = working_df["text"].str.upper()
+
+    #: remove periods, parenthesis, double colons, and slashes
+    logging.info("Removing periods, parenthesis, double colons, and slashes")
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace(".", "").strip(), axis=1)
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace("(", "").replace("(", "").strip(), axis=1)
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace(")", "").replace(")", "").strip(), axis=1)
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace("/", "").strip(), axis=1)
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace("\\", "").strip(), axis=1)
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace("::", ":").strip(), axis=1)
+
+    #: remove leading colons
+    logging.info("Removing leading colons")
+    working_df["text"] = working_df["text"].apply(lambda x: x.lstrip(":"))
+
+    #: replace euro symbol with 'E'
+    logging.info("Replacing euro symbol with 'E'")
+    working_df["text"] = working_df.apply(lambda r: r["text"].replace("€", "E").strip(), axis=1)
+
+    #: replaced accented characters with non-accented characters
+    logging.info("Replacing accented characters with non-accented characters")
+    working_df["text"] = working_df.apply(
+        lambda r: "".join(c for c in unicodedata.normalize("NFD", r["text"]) if unicodedata.category(c) != "Mn"), axis=1
+    )
+
+    #: flag parcels that contains special characters
+    working_df["special_char"] = "pass"
+    mask = working_df["text"].str.contains(r"[^\w:]") == True
+    working_df.loc[mask, "keep"] = "no"
+    working_df.loc[mask, "special_char"] = "fail"
+    logging.info("Number parcels with special characters flagged: %i", mask.value_counts()[1])
+
+    #: flag parcels that start with a letter or non-digit, should start with a number
+    working_df["nondigit_start"] = "pass"
+    mask = ~working_df["text"].str.contains(r"^\d") == True
+    working_df.loc[mask, "keep"] = "no"
+    working_df.loc[mask, "nondigit_start"] = "fail"
+    logging.info("Number parcels starting with a non-digit flagged: %i", mask.value_counts()[1])
+
+    #: flag parcels with a ':P' pattern; permits, not relevant - but PUE is okay!
+    working_df["permit"] = "pass"
+    mask = working_df["text"].str.contains(r"(?!:PUE):P") == True
+    working_df.loc[mask, "keep"] = "no"
+    working_df.loc[mask, "permit"] = "fail"
+    logging.info("Number of permits (:P pattern, but not :PUE) flagged: %i", mask.value_counts()[1])
+
+    #: flag parcels with a colon, that do not have a number before the colon
+    working_df["no_number_before_colon"] = "pass"
+    mask = working_df["text"].str.contains(r"^[^0-9]*:") == True
```

Comment:
```suggestion
    mask = working_df["text"].str.contains(r"^\D*:") == True
```

`\D` is any non digit so it's basically ^0-9

Reply:
Yeah, that's the same.